Fixed read_csv with CategoricalDtype with boolean categories (20498)

TomAugspurger opened the pull request:
- [x] closes #20498
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

TomAugspurger commented on pandas/core/arrays/categorical.py:
```diff
@@ -523,6 +524,13 @@ def _from_inferred_categories(cls, inferred_categories, inferred_codes,
                 cats = to_timedelta(inferred_categories, errors='coerce')
+            elif dtype.categories.is_boolean():
+                # from _true_values in parsers.pyx
+                # TODO: share declaration with parsers.pyx
```
Cython experts, is there a way to share a declaration with a non cython file?

We have

```
cdef object _true_values = [b'True', b'TRUE', b'true']
```

changing that do a `cpdef` didn't work for some reason.

(we need this, since the PythonParser's `true_values` can apparently be none.

jreback replied:
you need to remove the ``cdef`` then its importable just like anything else

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -523,6 +527,11 @@ def _from_inferred_categories(cls, inferred_categories, inferred_codes,
                 cats = to_datetime(inferred_categories, errors='coerce')
             elif is_timedelta64_dtype(dtype.categories):
                 cats = to_timedelta(inferred_categories, errors='coerce')
+            elif dtype.categories.is_boolean():
```
this does NOT belong here at all, you are mixing up parsing with setting the categories. you can do this via a separate function *inside* parser.pyx

TomAugspurger replied:
I think there were issues with that (either nogil, or having to share code with the Python parser).

jreback replied:
well it certainly doesn’t belong here
this is conflating parsing with categories

TomAugspurger replied:
I see now that this is called outside a nogil block, so a standalone method for that should be fine. If you want to open an issue though, because I don't think it should hold up the release.

jreback replied:
this isn't necessary for the RC. this needs to be fixed in this PR.

TomAugspurger commented:
I don't think I'll have time before the RC, and we're simply fixing a bug
in the existing code.

I'd like to keep the just merge bug fixes with the RC before the final, so
your call if you want to push to 0.23.1.

On Fri, Apr 27, 2018 at 6:56 AM, Jeff Reback <notifications@github.com>
wrote:

> *@jreback* commented on this pull request.
> ------------------------------
>
> In pandas/core/arrays/categorical.py
> <https://github.com/pandas-dev/pandas/pull/20826#discussion_r184666226>:
>
> > @@ -523,6 +527,11 @@ def _from_inferred_categories(cls, inferred_categories, inferred_codes,
>                  cats = to_datetime(inferred_categories, errors='coerce')
>              elif is_timedelta64_dtype(dtype.categories):
>                  cats = to_timedelta(inferred_categories, errors='coerce')
> +            elif dtype.categories.is_boolean():
>
> this isn't necessary for the RC. this needs to be fixed in this PR.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/20826#discussion_r184666226>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIilori4wAlmJJnIwNQm7XMGDPaHpks5tswdqgaJpZM4TkOVV>
> .
>

jreback commented:
then push to 0.23.1 (alternatively as @jorisvandenbossche mentoined, can make a 0.23rc milestone tag)

jreback commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -1155,6 +1155,7 @@ I/O
 - Bug in :func:`read_csv` where a ``MultiIndex`` with duplicate columns was not being mangled appropriately (:issue:`18062`)
 - Bug in :func:`read_csv` where missing values were not being handled properly when ``keep_default_na=False`` with dictionary ``na_values`` (:issue:`19227`)
 - Bug in :func:`read_csv` causing heap corruption on 32-bit, big-endian architectures (:issue:`20785`)
```
can you move to 0.23.1

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -503,12 +503,16 @@ def _from_inferred_categories(cls, inferred_categories, inferred_codes,
         dtype : CategoricalDtype or 'category'
+        true_values : list of bytes, optional
+            Uses the default `true_values` defined in parsers.pyx
+            by default.
```
versionadded?

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -523,6 +527,11 @@ def _from_inferred_categories(cls, inferred_categories, inferred_codes,
                 cats = to_timedelta(inferred_categories, errors='coerce')
+            elif dtype.categories.is_boolean():
+                if true_values is None:
+                    true_values = _true_values
```
you could *maybe* enhance ``to_numeric`` to take a ``downcasts='bool'`` to do this?

jorisvandenbossche commented:
@TomAugspurger do you have time to look at this one? (I don't know how far off it is)

TomAugspurger commented:
Possibly. If you can provide a unit test I can include it or you can make a followup PR including it.

jreback commented on pandas/tests/categorical/test_constructors.py:
```diff
@@ -468,23 +469,23 @@ def test_from_codes_with_categorical_categories(self):
     def test_from_inferred_categories(self, dtype):
         cats = ['a', 'b']
         codes = np.array([0, 0, 1, 1], dtype='i8')
-        result = Categorical._from_inferred_categories(cats, codes, dtype)
```
would be ok with removing these tests (as I think you have them covered directly in the parser)

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -370,6 +370,7 @@ MultiIndex
 I/O
 ^^^
 
+- Bug in :func:`read_csv` with a ``CategoricalDtype`` with boolean categories not correctly coercing the string values to booleans (:issue:`20498`)
```
is this enough from a user POV? need a small example?

jreback commented:
IIRC this was fine, can you rebase

pep8speaks commented:
Hello @TomAugspurger! Thanks for updating the PR.

 - There are no PEP8 issues in the file [`pandas/core/arrays/categorical.py`](https://github.com/TomAugspurger/pandas/blob/a1f64ff96ef2a4e9f84cad10733e073a5ad99e1c/pandas/core/arrays/categorical.py) !

 - There are no PEP8 issues in the file [`pandas/io/parsers.py`](https://github.com/TomAugspurger/pandas/blob/a1f64ff96ef2a4e9f84cad10733e073a5ad99e1c/pandas/io/parsers.py) !

 - There are no PEP8 issues in the file [`pandas/tests/arrays/categorical/test_constructors.py`](https://github.com/TomAugspurger/pandas/blob/a1f64ff96ef2a4e9f84cad10733e073a5ad99e1c/pandas/tests/arrays/categorical/test_constructors.py) !

 - There are no PEP8 issues in the file [`pandas/tests/io/parser/dtypes.py`](https://github.com/TomAugspurger/pandas/blob/a1f64ff96ef2a4e9f84cad10733e073a5ad99e1c/pandas/tests/io/parser/dtypes.py) !

jreback commented:
this looks good; haven’t looked at what is failing

jreback commented:
@TomAugspurger can you merge master

jreback commented:
@gfyoung if you'd like to rebase this, maybe can get this in

gfyoung commented:
Yikes!  Looks like this one got lost in the myriad of other PR's.  Sure thing.

gfyoung committed:
BUG: Properly handle CSV boolean CategoricalDtype

Previously, was being parsed as object instead of boolean.

Closes gh-20498.

Original Author: @TomAugspurger
Rebased by @gfyoung due to merge conflicts.

gfyoung commented:
@jreback  @TomAugspurger : Rebased this PR.  Wanted to note a couple of things (also helps to address many of the outstanding comments that were still current):

* I personally was not a fan of moving an entire `Categorical` method into `parsers.pyx`.  That IMO seems to do the kind of conflating that was pushed back against earlier.  I was able to get things to work without making this migration.  At the same time, the "parsing" is relegated to `parsers.pyx` before being passed to `Categorical._from_inferred_categories`.
* I'm okay with having a single `whatsnew` entry line.  I reworded it a little to make things clearer.
* Not sure we need a version added on the docstring, given that `Categorical._from_inferred_categories` is a private method by naming (starts with an underscore).

jreback commented:
looks ok to me. @TomAugspurger if you'd have a look.

gfyoung commented:
@TomAugspurger : Any thoughts on this?  I think we're waiting on your review before merging.

TomAugspurger commented:
Sorry for the delay, thanks for rebasing and fixing things up.
